# C3POa

C3POa (**C**oncatemeric **C**onsensus **C**aller with **P**artial **O**rder **a**lignments) is a computational pipeline for calling consensi on R2C2 nanopore data.

**This version of C3POa uses a new aligner (gonk). The old version of C3POa that uses water can be found [here](https://github.com/rvolden/C3POa/tree/water).**

## Dependencies

- [Python 3.6](https://www.python.org/downloads/)
- [NumPy 1.13.3](https://scipy.org/install.html)
- [poa v1.0.0 Revision: 1.2.2.9](https://github.com/tanghaibao/bio-pipeline)
- [Go](https://golang.org/dl/)
- [gonk](https://github.com/rvolden/gonk)
- [minimap2 2.7-r654](https://github.com/lh3/minimap2)
- [racon](https://github.com/isovic/racon)
- [blat source](https://users.soe.ucsc.edu/~kent/src/blatSrc35.zip) or [blat executable](http://hgdownload.soe.ucsc.edu/admin/exe/)

To fetch and build dependencies, use setup.sh.<br>
setup.sh will download and make the packages that you need to run C3POa (except Python, NumPy, Go, and blat).<br>
You don't need to have these in your PATH, but if you don't, you'll need to use a [config file](example_config). The setup script **does not** install programs or add them to your path. If you use the setup script, you still need to put the paths into a config file.

```bash
chmod +x setup.sh
./setup.sh
```

To install NumPy, you can go [here](https://scipy.org/install.html).<br>
Otherwise, you can use your computer's package manager (apt-get, dnf, brew, etc.) to install.<br>
Pip3 is another option for NumPy installation.<br>
Example:

```bash
sudo dnf install python3-numpy
```

or

```bash
pip3 install numpy
```

For blat, there are a couple options. You can build from [source](https://users.soe.ucsc.edu/~kent/src/blatSrc35.zip) or you can get an [executable](http://hgdownload.soe.ucsc.edu/admin/exe/). Please follow the documentation in the blat readme for make instructions.

--------------------------------------------------------------------------------

## Usage

After resolving all of the dependencies, you can run C3POa with python.

## C3POa_preprocessing.py

Takes raw 1D nanopore R2C2 reads in fastq format, removes low quality and short reads and then finds splint sequences in those reads using BLAT. It then adds the position of the splint to the read name and generates separate folders each containing a fastq file containing 4000 raw reads. Also demultiplexes reads based on splints that are put into the splint fasta file. You should end up with a directory that looks like: `somePath/number/Splint_1/R2C2_raw_reads.fastq` where Splint_1 would be which Splint that particular read was classified with.

Options (All required):

```
  -i  raw reads in fastq format

  -o  path where a splint_reads folder with the output files will be generated

  -q  only reads above this average quality will be retained (9 is recommended)

  -l  only reads longer than this number will be retained (1000 recommended)

  -s  sequence of DNA splint used in R2C2 protocol in fasta format

  -c  config file containing path to BLAT binary
```

```bash
python3 C3POa_preprocessing.py -i raw_reads.fastq -o output_path -q quality_cutoff
                               -l read_length_cutoff -s Splint_sequence.fasta
```

Example input read:

```
@63f115bc-6a91-42bd-a78a-667fd8255069
ACAGTCGATCATAGCTTAGCATGCATCGACGATCGATCGATCGA
+
"01&%"."I;"CSA"qr{X"uvc"\n"ggZ"Swj"yq"{wD"{z
```

Example output read (10 would be a splint position):

```
@63f115bc-6a91-42bd-a78a-667fd8255069_10
ACAGTCGATCATAGCTTAGCATGCATCGACGATCGATCGATCGA
+
"01&%"."I;"CSA"qr{X"uvc"\n"ggZ"Swj"yq"{wD"{z
```

--------------------------------------------------------------------------------

## C3POa.py

Takes fastq output produced by C3POa_preprocessing.py and generates consensus sequences in fasta format and a subread sequences in fastq format.

Options:

```
  -p  directory to which all temporary files will be written. Defaults to your current directory.

  -m  path to NUC.4.4.mat file (included in repository)

  -l  raw sequence length cutoff. Defaults to 1000

  -d  median distance between peaks cutoff. This should be the length of your shortest
      input sequence in your library preparation. Defaults to 500

  -c  config file containing paths to poa, racon, gonk, blat, and minimap2.

  -o  name (including path) of fasta file that the consensus gets written to. Defaults
      to R2C2_Consensus.fasta

  -z  use to exclude zero repeat reads

  -r  fastq file that contains reads generated by C3POa_preprocessing.py

  -t  use to print how long each dependency takes to run
```

```bash
python3 C3POa.py -r preprocessed_reads.fastq -p outpath -m path/to/NUC.4.4.mat -l 1000
                 -d 500 -c /path/to/config_file -o /path/to/consensus.fasta
```

Timing things and excluding zero repeat reads:

```bash
python3 C3POa.py -t -z -r preprocessed_reads.fastq -p outpath -m path/to/NUC.4.4.mat -l 1000
                 -d 500 -c /path/to/config_file -o /path/to/consensus.fasta
```

When you include -t (--timer), gonk, poa, racon, and consensus.py will be timed (times are directed to stdout).

When you include -z (--zero), C3POa will exclude zero repeat reads.

Example output read (readName_averageQuality_originalReadLength_numberOfRepeats_subreadLength):

```
>efbfbf09-7e2b-48e6-8e57-b3d36886739c_46.53_5798_2_1844
ACAGTCGATCATAGCTTAGCATGCATCGACGATCGATCGATCGA...
```

--------------------------------------------------------------------------------

## C3POa_postprocessing.py

Trims and reorients consensus sequences generated by C3POa.py to 5'->3' direction

```
  -i  fasta file containing consensus sequences generated by C3POa.py

  -o  directory which output files will be written to

  -a  sequence of cDNA adapter sequences in fasta format. Sequence names must be
      3Prime_adapter and 5Prime_adapter

  -c  config file containing path to BLAT binary

  -u  use to ignore read directionality

  -t  use to trim adapters off of the ends of the sequences
```

```bash
python3 C3POa_postprocessing.py -i /path/to/consensus.fasta -o out_path
                                -c /path/to/config_file -a /path/to/adapter.fasta
```
